eclair-core/src/main/scala/fr/acinq/eclair/db/jdbc/JdbcUtils.scala

```diff
@@ -96,20 +98,25 @@ trait JdbcUtils {
     }
   }
 
-  /**
-   * This helper assumes that there is a "data" column available, decodable with the provided codec
-   *
-   * TODO: we should use an scala.Iterator instead
-   */
-  def codecSequence[T](rs: ResultSet, codec: Codec[T]): Seq[T] = {
-    var q: Queue[T] = Queue()
-    while (rs.next()) {
-      q = q :+ codec.decode(BitVector(rs.getBytes("data"))).require.value
+  case class ExtendedResultSet(rs: ResultSet) extends Iterable[ResultSet] {
+
+    /**
+     * Iterates over all rows of a result set.
+     *
+     * Careful: the iterator is lazy, it must be materialized before the [[ResultSet]] is closed, by converting the end
+     * result in a collection or an option.
+     */
+    override def iterator: Iterator[ResultSet] = {
+      // @formatter:off
+      new Iterator[ResultSet] {
+        def hasNext: Boolean = rs.next()
```

Comment:
Is it ok that `hasNext` actually has a side-effect?
This only works if the calling pattern is always:
```scala
while (it.hasNext) it.next()
```

Reply:
Yes it works, because we always use this pattern to iterate once. Attempting to iterate twice on the same result set will yield no result on the second iteration, but that was already the case before this change.

However the abstraction does make it easier to shoot yourself in the foot because it hides away this kind of details.

Reply:
Our tests should catch new code that doesn't comply with this pattern, so we're probably ok.

Reply:
To be clear, this change really is a trade-off. It makes some things safer but can be riskier in other ways, and I'm not sure our tests will catch everything.

Reply:
🤞